Mathlib/Combinatorics/SetFamily/CauchyDavenport.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2023 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import Mathlib.Combinatorics.Additive.ETransform
+import Mathlib.GroupTheory.Order.Min
+
+/-!
+# The Cauchy-Davenport theorem
+
+This file proves a generalisation of the Cauchy-Davenport theorem to arbitrary groups.
+
+Cauchy-Davenport provides a lower bound on the size of `s + t` in terms of the sizes of `s` and `t`,
+where `s` and `t` are nonempty finite sets in a monoid. Precisely, it says that
+`|s + t| ≥ |s| + |t| - 1` unless the RHS is bigger than the size of the smallest nontrivial subgroup
+(in which case taking `s` and `t` to be that subgroup would yield a counterexample). The motivating
+example is `s = {0, ..., m}`, `t = {0, ..., n}` in the integers, which gives
+`s + t = {0, ..., m + n}` and `|s + t| = m + n + 1 = |s| + |t| - 1`.
+
+There are two kinds of proof of Cauchy-Davenport:
+* The first one works in linear orders by writing `a₁ < ... < aₖ` the elements of `s`,
+  `b₁ < ... < bₗ` the elements of `t`, and arguing that `a₁ + b₁ < ... < aₖ + b₁ < ... < aₖ + bₗ`
+  are distinct elements of `s + t`.
+* The second one works in groups by performing an "e-transform". Replace `s` and `t` by `s ∩ g • s`
+  and `t ∪ g⁻¹ • t`. For a suitably chosen `g`, this decreases `|s + t|` and keeps `|s| + |t|` the
+  same.
+
+## Main declarations
+
+* `Finset.min_le_card_mul`: A generalisation of the Cauchy-Davenport theorem to arbitrary groups.
+* `Monoid.IsTorsionFree.card_add_card_sub_one_le_card_mul`: The Cauchy-Davenport theorem in
+  torsion-free groups.
+* `ZMod.min_le_card_add`: The Cauchy-Davenport theorem.
+* `Finset.card_add_card_sub_one_le_card_mul`: The Cauchy-Davenport theorem in linear ordered
+  cancellative semigroups.
+
+## TODO
+
+Version for `circle`.
+
+## References
+
+* Matt DeVos, *On a generalization of the Cauchy-Davenport theorem*
+
+## Tags
+
+additive combinatorics, number theory, sumset, cauchy-davenport
+-/
+
+open Finset Function Monoid MulOpposite Subgroup
+open scoped Pointwise
+
+variable {α : Type*}
+
+/-! ### General case -/
+
+section General
+variable [Group α] [DecidableEq α] {x y : Finset α × Finset α} {s t : Finset α}
+
+/-- The relation we induct along in the proof of Cauchy-Davenport theorem. `(s₁, t₁) < (s₂, t₂)` iff
+* `|s₁ * t₁| < |s₂ * t₂|`
+* or `|s₁ * t₁| = |s₂ * t₂|` and `|s₂| + |t₂| < |s₁| + |t₁|`
+* or `|s₁ * t₁| = |s₂ * t₂|` and `|s₁| + |t₁| = |s₂| + |t₂|` and `|s₁| < |s₂|`. -/
+@[to_additive
+"The relation we induct along in the proof of Cauchy-Davenport theorem. `(s₁, t₁) < (s₂, t₂)` iff
+* `|s₁ + t₁| < |s₂ + t₂|`
+* or `|s₁ + t₁| = |s₂ + t₂|` and `|s₂| + |t₂| < |s₁| + |t₁|`
+* or `|s₁ + t₁| = |s₂ + t₂|` and `|s₁| + |t₁| = |s₂| + |t₂|` and `|s₁| < |s₂|`."]
+private def DevosMulRel : Finset α × Finset α → Finset α × Finset α → Prop :=
```

Comment:
This relation is internal to the proof and really uninteresting. In fact, I need this relation solely because I'm following a very general proof. No other proof uses this relation.

I could inline it at the cost of having to state the lemmas as `have` statements. At the very extreme, it could be a `termination_by` call instead.

The only purpose of having it be separated is for ease of writing and clarity. But it doesn't deserve to be part of the public API.

Reply:
My problem is that you have non-private lemmas talking about a private def, which can lead to weird situations in other files. Either make all of them private, or all of them non-private. 

Can you add in the docstring that this relation is used in the proof of Cauchy-Davenport by a guy named Devos, to clarify the name?

Reply:
Ah I see! Be reassured, all lemmas mentioning a `private` declaration in their statement are automatically marked `private` themselves.

Reply:
... Nevermind, it was the case in Lean 3 but this seems to be gone in Lean 4?!